src/main/java/org/alliancegenome/curation_api/services/DataProviderService.java

```diff
@@ -44,19 +57,100 @@ public DataProvider createAffiliatedModDataProvider() {
+		// find associated gene
+		Long geneID = getAssociatedGeneId(referencedCurie, entity.getSourceOrganization());
+		// if no gene found skip (= don't import) the accession
+		if (geneID == null) {
+			return new ObjectResponse<>();
+		}
+
+		DataProvider dbEntity = getDataProvider(entity.getSourceOrganization(), referencedCurie, entity.getCrossReference().getResourceDescriptorPage());
+		// we only create new records, no updates
+		if (dbEntity == null) {
+			dataProviderDAO.persist(entity);
+			if (!List.of("FB", "SGD").contains(entity.getSourceOrganization().getAbbreviation())) {
+				crossReferenceDAO.persistAccessionGeneAssociated(entity.getCrossReference().getId(), geneID);
+			}
+			return new ObjectResponse<>(entity);
+		}
+		return new ObjectResponse<>(dbEntity);
+	}
+
+	@NotNull
+	public static String getFullReferencedCurie(String localReferencedCurie) {
+		return RESOURCE_DESCRIPTOR_PREFIX + ":" + localReferencedCurie;
+	}
+
+	Map<String, Long> accessionGeneMap = new HashMap<>();
+	public static final String RESOURCE_DESCRIPTOR_PREFIX = "ENSEMBL";
+	public static final String RESOURCE_DESCRIPTOR_PAGE_NAME = "default";
+
+	private Long getAssociatedGeneId(String fullReferencedCurie, Organization sourceOrganization) {
+		String orgAbbreviation = sourceOrganization.getAbbreviation();
+		if (orgAbbreviation.equals("FB")) {
+			fullReferencedCurie = orgAbbreviation + ":" + fullReferencedCurie;
+		}
+		if (accessionGeneMap.size() == 0) {
+			if (List.of("FB", "SGD").contains(orgAbbreviation)) {
+				Map<String, Object> map = new HashMap<>();
+				map.put("displayName", orgAbbreviation);
+				Species species = speciesDAO.findByParams(map).getSingleResult();
+				accessionGeneMap = geneDAO.getAllGeneIdsPerSpecies(species);
+			} else {
+				ResourceDescriptorPage page = resourceDescriptorPageService.getPageForResourceDescriptor(RESOURCE_DESCRIPTOR_PREFIX, RESOURCE_DESCRIPTOR_PAGE_NAME);
+				accessionGeneMap = crossReferenceDAO.getGenesWithCrossRefs(page);
+			}
+		}
+		return accessionGeneMap.get(fullReferencedCurie);
+	}
+
+	// <crossReference.referencedCurie, DataProvider>
+	HashMap<String, DataProvider> dataProviderMap = new HashMap<>();
+
+	private DataProvider getDataProvider(Organization sourceOrganization, String crossReferenceCurie, ResourceDescriptorPage page) {
+		if (dataProviderMap.size() > 0) {
+			return dataProviderMap.get(crossReferenceCurie);
+		}
+		populateDataProviderMap(sourceOrganization, page);
+		return dataProviderMap.get(crossReferenceCurie);
+	}
+
+	private void populateDataProviderMap(Organization sourceOrganization, ResourceDescriptorPage page) {
+		HashMap<String, Object> params = new HashMap<>();
+		params.put("sourceOrganization.abbreviation", sourceOrganization.getAbbreviation());
+		params.put("crossReference.resourceDescriptorPage.name", page.getName());
+		List<DataProvider> allOrgProvider = dataProviderDAO.getAllDataProvider(params);
```

Comment:
Do we really need another method here when the findByParams can be called directly?

Reply:
Yes, as I want to get all records in a single query, ie I don't want to paginate. All the logic is encapsulated in that new method.